File: 2024/6/main.py
import aiofiles
import asyncio
import numpy as np

from enum import Enum
import time

from tqdm import tqdm

class Direction(Enum):
    NORTH = 1
    EAST  = 2
    SOUTH = 3
    WEST  = 4
    
    def next(self):
        if self == Direction.NORTH:
            return Direction.EAST
        elif self == Direction.EAST:
            return Direction.SOUTH
        elif self == Direction.SOUTH:
            return Direction.WEST
        elif self == Direction.WEST:
            return Direction.NORTH

dirs = {
    Direction.NORTH: (-1, 0),
    Direction.EAST:  (0, 1),
    Direction.SOUTH: (1, 0),
    Direction.WEST:  (0, -1)
}

def is_in_bounds(actor, shape):
    return actor[0] >= 0 and actor[0] < shape[0] and actor[1] >= 0 and actor[1] < shape[1]

async def main():
    puzzle = []
    orig_actor = (0, 0)
    orig_direction = Direction.NORTH
    
    async with aiofiles.open('input.txt', 'r') as file:
        async for line in file:
            row = []
            for char in line:
                if char == '.':
                    row.append(0)
                elif char == '#':
                    row.append(1)
                elif char == '^':
                    x = len(row)
                    y = len(puzzle)
                    row.append(0)
                    orig_actor = (y, x)
            
            puzzle.append(row)            

    puzzle = np.array(puzzle, dtype=np.byte)
    shape = puzzle.shape
    actor = orig_actor
    direction = orig_direction
    previous_positions = set()
    
    while True:
        previous_positions.add((actor[0], actor[1]))
        
        move = dirs[direction]
        next = (actor[0] + move[0], actor[1] + move[1])
        if not is_in_bounds(next, shape):
            break
        elif puzzle[next[0], next[1]] == 1:
            direction = direction.next()
        else:
            actor = next
        
    print(len(previous_positions))
    
    causes_loop = 0
    for idx in tqdm(previous_positions):
        actor = orig_actor
        direction = orig_direction
        directional_previous_position = set()
        
        while True:
            if (direction, actor[0], actor[1]) in directional_previous_position:
                causes_loop += 1
                break
            else:
                directional_previous_position.add((direction, actor[0], actor[1]))
            
            move = dirs[direction]
            next = (actor[0] + move[0], actor[1] + move[1])
            if not is_in_bounds(next, shape):
                break
            elif puzzle[next[0], next[1]] == 1:
                direction = direction.next()
            elif next == (idx[0], idx[1]):
                direction = direction.next()
            else:
                actor = next
            
    print(causes_loop)
    
if __name__=='__main__':
    t1 = time.time()
    asyncio.run(main())
    t2 = time.time()
    print(f"Execution time: {(t2-t1): .2f}s")

File: 2024/6/main.cpp
#include <iostream>
#include <fstream>
#include <unordered_set>
#include <chrono>

using namespace std;

enum Direction {
    NORTH,
    EAST,
    SOUTH,
    WEST
};

static int dirs[4][2] = {
    {-1, 0},
    {0, 1},
    {1, 0},
    {0, -1},
};

Direction operator++(Direction &dir) {
    dir = static_cast<Direction>((dir + 1) % 4);
    return dir;
}

bool is_inbounds(int x, int y, int n, int m) {
    return x >= 0 && x < n && y >= 0 && y < m;
}

int simple_hash(int x, int y, int d) {
    // x and y are at most 255 d is at most 4
    // 8 bits for x, 8 bits for y, 8 bits for d
    return (x << 16) | (y << 8) | d;
}
int simple_hash(int x, int y) {
    // x and y are at most 255 d is at most 4
    // 8 bits for x, 8 bits for y, 8 bits for d
    return (x << 8 | y);
}
void print_grid(char **grid, int grid_x, int grid_y) {
    for (int i = 0; i < grid_y; ++i) {
        for (int j = 0; j < grid_x; ++j) {
            cout << grid[i][j];
        }
        cout << endl;
    }
}

int main() {
    auto start = chrono::high_resolution_clock::now();

    // read file width & height
    ifstream fin("input.txt");
    string unused;
    getline(fin, unused);
    int grid_x = unused.length();
    
    int grid_y = 1;
    while (getline(fin, unused)) {
        ++grid_y;
    }

    // reset file
    fin.clear();
    fin.seekg(0, ios::beg);

    // load data into grid
    char **grid = new char*[grid_y];

    int orig_x, orig_y;
    char c;

    for (int i = 0; i < grid_y; ++i) {
        grid[i] = new char[grid_x];
        for (int j = 0; j < grid_x; ++j) {
            fin >> noskipws >> c;
            if (c == '\n') {
                fin >> noskipws >> c;
            }
            grid[i][j] = c;
            if (c == '^') {
                orig_x = j;
                orig_y = i;
            }
        }
    }
    // print_grid(grid, grid_x, grid_y);

    // follow the path
    Direction d;
    int x, y;
    int next_x, next_y;
    unordered_set<int> visited = unordered_set<int>();
    // set initial position
    d = NORTH;
    x = orig_x;
    y = orig_y;

    while (true) {
        visited.insert(simple_hash(x, y));
        
        next_x = x + dirs[d][1];
        next_y = y + dirs[d][0];

        if (!is_inbounds(next_x, next_y, grid_x, grid_y)) {
            break;
        } else if (grid[next_y][next_x] == '#') {
            ++d;
        } else {
            x = next_x;
            y = next_y;
        }
    }
    // print places visited
    cout << "Visited: " << visited.size() << endl;

    // iterate over visited
    int loops = 0;
    int block_x, block_y;
    int steps;
    unordered_set<int> directional_visited = unordered_set<int>();

    for (auto it = visited.begin(); it != visited.end(); ++it) {
        // reverse the hash teehee
        block_x = (*it) >> 8;
        block_y = (*it) & 0xFF;

        x = orig_x;
        y = orig_y;
        d = NORTH;
        directional_visited.clear();

        while (true) {
            next_x = x + dirs[d][1];
            next_y = y + dirs[d][0];

            if (!is_inbounds(next_x, next_y, grid_x, grid_y)) {
                break;
            } else if (grid[next_y][next_x] == '#') {
                ++d;
                if (directional_visited.contains(simple_hash(x, y, d))) {
                    ++loops;
                    break;
                }
                directional_visited.insert(simple_hash(x, y, d));
            } else if (next_x == block_x && next_y == block_y) {
                ++d;
                if (directional_visited.contains(simple_hash(x, y, d))) {
                    ++loops;
                    break;
                }
                directional_visited.insert(simple_hash(x, y, d));
            } else {
                x = next_x;
                y = next_y;
                ++steps;
            }
        }
    }
    cout << "Causes loops: " << loops << endl;

    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    cout << "Execution time: " << duration.count() << "ms" << endl;

    for (int i = 0; i < grid_y; ++i) {
        delete[] grid[i];
    }
    delete[] grid;

    return 0;
}
